core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -13,105 +13,123 @@
+import static io.trino.spi.predicate.TupleDomain.columnWiseUnion;
 import static java.util.Objects.requireNonNull;
 import static java.util.function.Function.identity;
 
 public class LocalDynamicFilterConsumer
         implements DynamicFilterSourceConsumer
 {
-    private static final int PARTITION_COUNT_INITIAL_VALUE = -1;
     // Mapping from dynamic filter ID to its build channel indices.
     private final Map<DynamicFilterId, Integer> buildChannels;
-
     // Mapping from dynamic filter ID to its build channel type.
     private final Map<DynamicFilterId, Type> filterBuildTypes;
-
     private final List<Consumer<Map<DynamicFilterId, Domain>>> collectors;
+    private final long domainSizeLimitInBytes;
 
     // Number of build-side partitions to be collected, must be provided by setPartitionCount
     @GuardedBy("this")
-    private int expectedPartitionCount = PARTITION_COUNT_INITIAL_VALUE;
-
+    private Integer expectedPartitionCount;
     @GuardedBy("this")
-    private boolean collected;
-
-    // The resulting predicates from each build-side partition.
-    @Nullable
+    private int collectedPartitionCount;
     @GuardedBy("this")
-    private List<TupleDomain<DynamicFilterId>> partitions;
+    private volatile boolean collected;
 
-    public LocalDynamicFilterConsumer(
-            Map<DynamicFilterId, Integer> buildChannels,
-            Map<DynamicFilterId, Type> filterBuildTypes,
-            List<Consumer<Map<DynamicFilterId, Domain>>> collectors)
+    private final Queue<TupleDomain<DynamicFilterId>> summaryDomains = new ConcurrentLinkedQueue<>();
+
+    public LocalDynamicFilterConsumer(Map<DynamicFilterId, Integer> buildChannels, Map<DynamicFilterId, Type> filterBuildTypes, List<Consumer<Map<DynamicFilterId, Domain>>> collectors, DataSize domainSizeLimit)
     {
         this.buildChannels = requireNonNull(buildChannels, "buildChannels is null");
         this.filterBuildTypes = requireNonNull(filterBuildTypes, "filterBuildTypes is null");
         verify(buildChannels.keySet().equals(filterBuildTypes.keySet()), "filterBuildTypes and buildChannels must have same keys");
-
         requireNonNull(collectors, "collectors is null");
         checkArgument(!collectors.isEmpty(), "collectors is empty");
-        this.collectors = collectors;
-        this.partitions = new ArrayList<>();
+        this.collectors = ImmutableList.copyOf(collectors);
+        this.domainSizeLimitInBytes = requireNonNull(domainSizeLimit, "domainSizeLimit is null").toBytes();
     }
 
     @Override
-    public void addPartition(TupleDomain<DynamicFilterId> tupleDomain)
+    public void addPartition(TupleDomain<DynamicFilterId> domain)
     {
+        if (collected) {
+            return;
+        }
+
+        summaryDomains.add(domain);
+        unionSummaryDomains();
+
         TupleDomain<DynamicFilterId> result;
         synchronized (this) {
+            verify(expectedPartitionCount == null || collectedPartitionCount < expectedPartitionCount);
```

Comment:
can we make `synchronized` block wider and check `if(collected)` only once?
Possibly other things could be simplified to. If we did `unionSummaryDomains` always under lock then 
`// summary can be null as another concurrent summary compaction may be running` would no longer be true and code would be simpler.

Reply:
Also runing `final` compaction would not be needed I think.
Are there reasons for making it more complex than it seems necessary?

Reply:
The union operation can be fairly expensive and operators (for example the filters collection operator before HashBuilderOperator) tend to finish all at the same time what could result in up to `32` threads being blocked waiting for all the unions to finish.

Reply:
My guess it is not needed. But if you insist please comment code so it is obvious for the reader why the design choice.